rfcs/text/0408-versioning-types.md

```diff
@@ -0,0 +1,355 @@
+
+#### Data Type Reference
+
+A data type reference is an inline reference to a data-type. A data type reference will be compatible with another data type reference if the data types that are referred to are also compatible.
+
+> _Without_ non-primitive data types this implies that the references are to the same data type.
+
+#### Property Type Object
+
+A poperty type object is JSON object defining a schema which is used as `oneOf` the permitted values for a property type, where the schema is of `type: "object`, and within its own `properties` definition:
+
+1.  The keys are base URI's to a property type
+1.  The values are defined by either:
+    1.  A reference to a property type
+    1.  An array definition, where
+        1.  the `items` are defined by a reference to a property type
+        1.  `minItems` and/or `maxItems` are optionally defined
+1.  The required fields are defined by a `required` list where the elements are base URIs that are a subset of the keys
+
+A property type Object `X` is therefore compatible with another property type Object `Y` if and only if
+
+1.  Each key in the `properties` of `X` is also in the `properties` of `Y`
+1.  If a property in `X` is defined as an array, then it's also defined as an array in `Y` (and therefore the same for direct references)
+1.  If a property in `X` is defined as an array with a `minItems`, then there is a `minItems` constraint on the respective property in `Y` that is equal to or larger to that in `X`
+1.  If a property in `X` is defined as an array with a `maxItems`, then there is a `maxItems` constraint on the respective property in `Y` that is less than or equal to that in `X`
+1.  Each property type referenced in `X` is compatible with the respective property type referenced in `Y`. Where respective is defined by having the same base URI (property key)
+1.  Each URI in the `required` list of `X` is also in the `required` list of `Y`
+
+#### An Array of "Property Values"
+
+Finally, the "Property Values" can be a recursive definition that infinitely nests arrays where the `items` are a `oneOf` where the options are "Property Values" variants. Because of this, compatibility is defined largely similar to above.
+
+A definition of an array of "Property Values" `X` is compatible with another definition of an array of "Property Values" `Y`, if and only if
+
+1.  Every element of `X`'s `oneOf` is compatible with at least one element of `Y`'s `oneOf`
+1.  If `X` has a `minItems` constraint specified, then there is a `minItems` constraint in `Y` that is equal to or larger to that in `X`
+1.  If `X` has a `maxItems` constraint specified, then there is a `maxItems` constraint in `Y` that is less than or equal to that in `X`
+
+### Entity Types
+
+1.  Entity types define their constraints on the structure of an entity through their `properties`, where
```

Comment:
`required` (for properties) is missing in this list

Reply:
Added, good spot